nifi-nar-bundles/nifi-asn1-bundle/nifi-asn1-services/src/main/java/org/apache/nifi/jasn1/JASN1Reader.java

```diff
@@ -0,0 +1,315 @@
+import org.apache.nifi.controller.ControllerServiceInitializationContext;
+import org.apache.nifi.expression.ExpressionLanguageScope;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.reporting.InitializationException;
+import org.apache.nifi.schema.access.SchemaNotFoundException;
+import org.apache.nifi.serialization.MalformedRecordException;
+import org.apache.nifi.serialization.RecordReader;
+import org.apache.nifi.serialization.RecordReaderFactory;
+import org.apache.nifi.util.file.classloader.ClassLoaderUtils;
+
+import javax.tools.JavaCompiler;
+import javax.tools.JavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import javax.tools.ToolProvider;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Tags({"asn", "ans1", "jasn.1", "jasn1", "record", "reader", "parser"})
+@CapabilityDescription("Reads ASN.1 content and creates NiFi records.")
+public class JASN1Reader extends AbstractConfigurableComponent implements RecordReaderFactory {
+
+    private static final PropertyDescriptor ROOT_MODEL_NAME = new PropertyDescriptor.Builder()
+        .name("root-model-name")
+        .displayName("Root Model Name")
+        .description("The model name in the form of 'MODULE-NAME.ModelType'. " +
+            "Mutually exclusive with and should be preferred to 'Root Model Class Name'. (See additional details for more information.)")
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .required(false)
+        .build();
+
+    private static final PropertyDescriptor ROOT_CLASS_NAME = new PropertyDescriptor.Builder()
+        .name("root-model-class-name")
+        .displayName("Root Model Class Name")
+        .description("A canonical class name that is generated by the ASN.1 compiler to encode the ASN.1 input data. Mutually exclusive with 'Root Model Name'." +
+            " Should be used when the former cannot be set properly. See additional details for more information.")
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .required(false)
+        .build();
+
+    /**
+     * Not included!
+     * To make this service as simple as possible, records are to be expected to correspond to a concrete ASN type.
+     * Not removing though, should it be required in the future.
+     */
+    private static final PropertyDescriptor RECORD_FIELD = new PropertyDescriptor.Builder()
```

Comment:
The problem is that the code is already referencing these according to Koji's original idea.

I removed it from the public interface because it was not needed so testing this would require time and effort that is not worth it really. I don't think it's possible to extract the relevant changes in a clean manner.
On the other hand, completely eradicating them from the codebase feels risky and resource-intensive. (Testing all this is really hard.) Also, should we need these concepts later, we would need to implement them from scratch.

All-in-all I would consider these as half-finished extension points. I know this is not ideal  but I chose this approach because all things considered this seemed to be the best compromise.